### PR TITLE
feat: statistics View with bar and line charts

### LINKED ISSUE #495: Statistics View - API and Usages
### Is your feature request related to a problem? Please describe

Currently, it's not possible to see statistics about the API and it's usages in the editor.

### Desired solution

Create a new view that visualizes statistics about the API, it's usages, and the annotations.

#### Data about the size of the API

* Total number of API elements
  * Classes
  * Functions
  * Parameters
* Number of public API elements (see above)
* Number of used API elements (see above)
* Number of useful parameters (set to at least two different values)

This should be displayed as a labeled bar chart, similar to this:

image

It should be possible to set different thresholds, so it's for example possible to consider API elements used only if they are used at least 3 times.

Finally, the distribution of remaining classes, functions, and parameters should be shown for different thresholds in a line chart, with the threshold on the x-axis and the number of remaining API elements on the y-axis.

min_usages_vs_classes
min_usages_vs_functions
min_usages_vs_params

### Additional context (optional)

* Blocked by #494.
* Relates to #487.
* https://react-d3-library.github.io/ can be used to create the charts.

### PR DESCRIPTION
**Pull request by nvollroth**
Closes #495 .

### Summary of Changes

See Issue.

### Testing Instructions

1. Start the application with `npm run dev` 
2. Import your api-data file
3. Click on `Statistics View` and watch your results in the right part of your screen

### CODE REVIEW

File: api-editor/gui/src/features/packageData/selectionView/StatisticsView.tsx [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+
+    result.push(parameters.length);
+    result.push(publicParameters);
+    result.push(getUsedParameters(usages, usedThreshold).length);
+    result.push(getUsefulParameters(usages, usedThreshold));
+
+    return result;
+};
+
+let getUsedClasses = function (usages: UsageCountStore, usedThreshold: number): number {
```
Comment (lars-reimann) [inline api-editor/gui/src/features/packageData/selectionView/StatisticsView.tsx:102]: Should only count public used classes.

File: api-editor/gui/src/features/packageData/selectionView/StatisticsView.tsx [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    let usedClasses = 0;
+    usages.classUsages.forEach((value) => {
+        usedClasses += value >= usedThreshold ? 1 : 0;
+    });
+    return usedClasses;
+};
+
+let getUsedFunctions = function (usages: UsageCountStore, usedThreshold: number): number {
```
Comment (lars-reimann) [inline api-editor/gui/src/features/packageData/selectionView/StatisticsView.tsx:110]: Should only count public used functions.

File: api-editor/gui/src/features/packageData/selectionView/StatisticsView.tsx (hunk 1 above)
Comment (lars-reimann) [reply api-editor/gui/src/features/packageData/selectionView/StatisticsView.tsx:102]: Also count move this to the `UsageCountStore`.

File: api-editor/gui/src/features/packageData/selectionView/StatisticsView.tsx (hunk 2 above)
Comment (lars-reimann) [reply api-editor/gui/src/features/packageData/selectionView/StatisticsView.tsx:110]: Also count move this to the `UsageCountStore`.

File: api-editor/gui/src/features/packageData/selectionView/StatisticsView.tsx [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    let usedFunctions = 0;
+    usages.functionUsages.forEach((value) => {
+        usedFunctions += value >= usedThreshold ? 1 : 0;
+    });
+    return usedFunctions;
+};
+
+let getUsedParameters = function (usages: UsageCountStore, usedThreshold: number): string[] {
```
[the remark this answers is not in the document]
Comment (lars-reimann) [reply api-editor/gui/src/features/packageData/selectionView/StatisticsView.tsx:118]: Also count move this to the `UsageCountStore`.

File: api-editor/gui/src/features/packageData/selectionView/StatisticsView.tsx [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    let usedParameters: string[] = [];
+    usages.parameterUsages.forEach((value, key) => {
+        if (value >= usedThreshold) {
+            usedParameters.push(key);
+        }
+    });
+    return usedParameters;
+};
+
+let getUsefulParameters = function (usages: UsageCountStore, usedThreshold: number): number {
```
[the remark this answers is not in the document]
Comment (lars-reimann) [reply api-editor/gui/src/features/packageData/selectionView/StatisticsView.tsx:128]: Also count move this to the `UsageCountStore`.

File: api-editor/gui/src/features/packageData/selectionView/StatisticsView.tsx [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+    const usedParameters = getUsedParameters(usages, usedThreshold);
+    let usefulParameter = 0;
+    usedParameters.forEach((name) => {
+        // @ts-ignore
+        if (usages.valueUsages.get(name).size > 1) {
```
[the remark this answers is not in the document]
Comment (lars-reimann) [reply api-editor/gui/src/features/packageData/selectionView/StatisticsView.tsx:133]: Also count move this to the `UsageCountStore`.

**Comment by lars-reimann** [comment]
I know this is still WIP, but I'd forget this otherwise until Friday:

There is currently a difference between the values in the bar charts and what the filters return:

|Filter string|Number of matched nodes|Number in bar|
|-|-|-|
|`is:class is:public usages:>0`|217|257 (used classes)|
|`is:function is:public usages:>0`|690|795 (used functions)|
|`is:parameter is:public usages:>0`|2136|2522 (used parameters)|
|`is:parameter is:public usefulness:>0`|1857|2043 (used classes)|

One set of numbers is wrong. The line chart might have similar problems but I did not look into this.

File: api-editor/gui/src/features/packageData/selectionView/StatisticsView.tsx (hunk 1 above)
Comment (GideonKoenig) [reply api-editor/gui/src/features/packageData/selectionView/StatisticsView.tsx]: done

File: api-editor/gui/src/features/packageData/selectionView/StatisticsView.tsx (hunk 2 above)
Comment (GideonKoenig) [reply api-editor/gui/src/features/packageData/selectionView/StatisticsView.tsx]: done

File: api-editor/gui/src/features/packageData/selectionView/StatisticsView.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (GideonKoenig) [reply api-editor/gui/src/features/packageData/selectionView/StatisticsView.tsx]: done

File: api-editor/gui/src/features/packageData/selectionView/StatisticsView.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (GideonKoenig) [reply api-editor/gui/src/features/packageData/selectionView/StatisticsView.tsx]: done

**Comment by GideonKoenig** [comment]
@lars-reimann ready for review

**Comment by lars-reimann** [comment]
I'll have a look at it tomorrow.

**Review by lars-reimann — APPROVED** [review]

**Pull request merged**